### ISSUE TITLE
Penalization term

### ISSUE THREAD

**Issue by moussaKam** [kaushalshetty/Structured-Self-Attention#12]
In the original paper, the penalization term is equal to the Frobenius norm squared. In your implementation you did not consider the power 2. Did you try comparing both cases? 
![frobenius](https://user-images.githubusercontent.com/28675016/74054048-7b181f80-49dd-11ea-9a0d-c3987ac2fd28.png)

**Comment by kaushalshetty** [comment]
The implementation is indeed l2 norm. Please refer the below line. I missed it out in the code comments.

https://github.com/kaushalshetty/Structured-Self-Attention/blob/84ed64f9cee75f84a5d871a46508a308182e4a6d/attention/model.py#L131

**Comment by moussaKam** [comment]
In the paper, the Frobenius norm is squared, while in your code it's not.

**Comment by kaushalshetty** [comment]
If you see the previous code line, I am taking the square root of the sum of squares. `(torch.sum(m**2)) `
This squares the input and sums it up.

**Comment by moussaKam** [comment]
Why do you even need the square root if the term is squared? Please refer to the equation in the paper. In your code you're implementing the Frobenius norm, not the Frobenius norm to the power 2. 
In other words, the penalization term in your code is: 
![frobenius](https://user-images.githubusercontent.com/28675016/74352943-81c9dc80-4db9-11ea-8317-68a34c6edef8.png)
Instead of: 
![frobenius_squared](https://user-images.githubusercontent.com/28675016/74352967-88585400-4db9-11ea-9309-f941cff91c37.png)

**Comment by kaushalshetty** [comment]
`Why do you even need the square root if the term is squared?`
Frobenius norm is defined as the square root of sum of squares of matrix. It is called a "norm" because we are normalizing the squares using square root. It will not cancel out the squared terms if you assumed so.

`Please refer to the equation in the paper. In your code you're implementing the Frobenius norm, not the Frobenius norm to the power 2.`
Yes, you are right. The implementation just uses a Frobenius Norm but the paper describes an unnormalized version of the Frobenius. I have not tested with the unnormalized norm but I would expect a stronger penalization due to its larger magnitude.

**Comment by moussaKam** [comment]
> It will not cancel out the squared terms if you assumed 

I didn't assume that. I didn't mean the matrix entries. I'm talking about squaring the Frobenius norm which would indeed cancel out the square root. 
Anyway thank you for your time!
